### ISSUE TITLE
Encountered a timeout while trying to join the network

### ISSUE THREAD

**Issue by Vort** [maidsafe/safe_network#2063]
During `feb2` testnet today I attempted to join network with my node.
Eventually, join succeeded, but first two tries were failures:
```text
Node started
(PID: 7924): Encountered a timeout while trying to join the network. Retrying after 30 seconds. Node log path: d:\SN
(PID: 7924): Encountered a timeout while trying to join the network. Retrying after 30 seconds. Node log path: d:\SN
Node PID: 7924, prefix: Prefix(1), name: a38f63(10100011).., age: 5, connection info: 
```

Ignoring node join looks like a bug. Maybe it can be solved by increasing priority of join messages processing?

This is how node logs looks like in cases of failure:
```text
[2023-02-02T16:50:51.528623Z INFO sn_node] 
Running sn_node v0.73.2
=======================
[2023-02-02T16:50:51.529623Z TRACE sn_node] Initial node config: Config { wallet_id: None, root_dir: Some("g:\\SN_Root"), verbose: 2, completions: None, log_dir: Some("d:\\SN"), logs_retained: 0, logs_max_bytes: 10485760, logs_max_lines: 0, logs_uncompressed: 100, update: false, update_only: false, json_logs: false, resource_logs: false, clear_data: false, first: false, network_contacts_file: Some("d:\\SN\\default4"), local_addr: None, skip_auto_port_forwarding: false }
[2023-02-02T16:50:51.625643Z INFO sn_node::node::node_starter] 682085.. Bootstrapping as a new node.
[2023-02-02T16:50:51.633644Z INFO sn_node::node::node_starter] 682085.. Starting a new node (PID: 7924) with socket: 0.0.0.0:61098, network's genesis key: PublicKey(13ec..912f)
[2023-02-02T16:50:51.634644Z INFO sn_node::node::node_starter] Node 682085.. started.
[2023-02-02T16:50:51.634644Z DEBUG sn_node::node::flow_ctrl] [NODE READ]: flowctrl node context lock got
[2023-02-02T16:50:51.635645Z TRACE sn_node::node::flow_ctrl::cmd_ctrl] Spawned process for cmd TryJoinNetwork, id: [0]
[2023-02-02T16:50:51.635645Z INFO sn_node::node::flow_ctrl::dispatcher] [NODE READ]: getting lock for try_join_section
[2023-02-02T16:50:51.635645Z INFO sn_node::node::flow_ctrl::dispatcher] [NODE READ]: got lock for try_join_section
[2023-02-02T16:50:51.635645Z TRACE sn_node::node::flow_ctrl::cmd_ctrl] Spawned process for cmd SendMsgEnqueueAnyResponse { msg: TryJoin(None), msg_id: MsgId(02db..f66b), recipients: {Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }} }, id: [0, 0]
[2023-02-02T16:50:51.635645Z DEBUG sn_node::node::messaging::streams] Sending out + awaiting response of MsgId(02db..f66b) to 1 holder node/s {Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }}
[2023-02-02T16:50:51.635645Z INFO sn_node::node::messaging::streams] About to send MsgId(02db..f66b) to holder node: Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }
[2023-02-02T16:50:51.761670Z DEBUG sn_node::node::messaging::streams] Response in from Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 } for MsgId(02db..f66b): WireMsg { header: WireMsgHeader { version: 1, msg_envelope: MsgEnvelope { msg_id: MsgId(02db..f66b), kind: Node { name: d392c8(11010011).., is_join: false, is_ae: true } } }, dst: Dst { name: 682085(01101000).., section_key: PublicKey(02d7..c705) } }
[2023-02-02T16:50:51.761670Z TRACE sn_node::node::flow_ctrl::cmd_ctrl] Spawned process for cmd HandleMsg { origin: Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }, wire_msg: WireMsg { header: WireMsgHeader { version: 1, msg_envelope: MsgEnvelope { msg_id: MsgId(02db..f66b), kind: Node { name: d392c8(11010011).., is_join: false, is_ae: true } } }, dst: Dst { name: 682085(01101000).., section_key: PublicKey(02d7..c705) } }, send_stream: None }, id: [0, 0, 0]
[2023-02-02T16:50:51.761670Z TRACE sn_node::node::messaging/handle_msg] Handling msg MsgId(02db..f66b). Checking for AE first...
[2023-02-02T16:50:51.761670Z TRACE sn_node::node::messaging/handle_msg] [NODE READ]: Handle msg read lock attempt success
[2023-02-02T16:50:51.780674Z TRACE sn_node::node::messaging::node_msgs/handle_msg] NodeMsgToBeHandled: MsgId(02db..f66b)
[2023-02-02T16:50:51.780674Z TRACE sn_node::node::messaging::node_msgs/handle_msg] Handling msg: AE from d392c8.. at 134.209.24.43:12000: MsgId(02db..f66b)
[2023-02-02T16:50:51.780674Z DEBUG sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] [NODE READ]: handling AE read gottt...
[2023-02-02T16:50:51.780674Z WARN sn_interface::network_knowledge::section_tree/handle_msg/handle_anti_entropy_msg] Could not find related section to Prefix(1) in order to validate SAP's section is not shrinking
[2023-02-02T16:50:51.820682Z INFO sn_interface::network_knowledge/handle_msg/handle_anti_entropy_msg] Updated network section tree with SAP for Prefix(1)
[2023-02-02T16:50:51.837685Z DEBUG sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] [NODE WRITE]: handling AE write gottt...
[2023-02-02T16:50:51.837685Z WARN sn_interface::network_knowledge::section_tree/handle_msg/handle_anti_entropy_msg] Could not find related section to Prefix(1) in order to validate SAP's section is not shrinking
[2023-02-02T16:50:51.870692Z INFO sn_interface::network_knowledge/handle_msg/handle_anti_entropy_msg] Updated network section tree with SAP for Prefix(1)
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] net knowledge udpated
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] updated for section change
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] [NODE READ] Latest context read
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] [NODE READ] Latest context got.
[2023-02-02T16:50:51.871692Z TRACE sn_node::node::messaging::update_section/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] DataReorganisationUnderway
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::messaging::update_section/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] Querying section for any new data
[2023-02-02T16:50:51.871692Z TRACE sn_node::storage::register_store/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] Listening all register addrs
[2023-02-02T16:50:51.871692Z TRACE sn_node::storage::register_store/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] Listening all register addrs done
[2023-02-02T16:50:51.871692Z TRACE sn_node::node::messaging::update_section/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] section neighbours for data req: 1: {Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }}
[2023-02-02T16:50:51.871692Z TRACE sn_node::node::messaging::update_section/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] Sending our data list to: {Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }}
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::flow_ctrl::cmds/handle_msg/handle_anti_entropy_msg/ask_for_any_new_data_from_whole_section] Sending msg MsgId(ae4f..5cc0) to Multiple({Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }}): NodeDataCmd(SendAnyMissingRelevantData([]))
[2023-02-02T16:50:51.871692Z INFO sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] SectionTree written to disk with update for prefix Prefix(1)
[2023-02-02T16:50:51.871692Z TRACE sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] AeResendAfterRetry
[2023-02-02T16:50:51.871692Z TRACE sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] Resend Original MsgId(02db..f66b) to Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 } with TryJoin(None)
[2023-02-02T16:50:51.871692Z TRACE sn_node::node::messaging::anti_entropy/handle_msg/handle_anti_entropy_msg] AeResendAfterRedirect
[2023-02-02T16:50:51.871692Z DEBUG sn_node::node::flow_ctrl::cmds/handle_msg/handle_anti_entropy_msg] Sending msg MsgId(59b8..124d) to Single(Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }): TryJoin(None)
[2023-02-02T16:50:51.872692Z TRACE sn_node::node::flow_ctrl::cmd_ctrl] Spawned process for cmd SendMsg { msg: TryJoin(None), msg_id: MsgId(59b8..124d), recipients: Single(Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }) }, id: [0, 0, 0, 1]
[2023-02-02T16:50:51.872692Z TRACE sn_node::node::messaging::node_msgs] Sending msg: MsgId(59b8..124d)
[2023-02-02T16:50:51.872692Z TRACE sn_node::node::flow_ctrl::cmd_ctrl] Spawned process for cmd SendMsg { msg: NodeDataCmd(SendAnyMissingRelevantData([])), msg_id: MsgId(ae4f..5cc0), recipients: Multiple({Peer { name: d392c8(11010011).., addr: 134.209.24.43:12000 }}) }, id: [0, 0, 0, 0]
[2023-02-02T16:50:51.872692Z TRACE sn_node::node::messaging::node_msgs] Sending msg: MsgId(ae4f..5cc0)
[2023-02-02T16:51:51.637809Z ERROR sn_node::node::node_starter] Failed join: Elapsed(())
[2023-02-02T16:51:51.637809Z ERROR sn_node] (PID: 7924): Encountered a timeout while trying to join the network. Retrying after 30 seconds.
[2023-02-02T16:51:51.637809Z WARN qp2p::connection] Connection handle dropped, thus closing it, conn_id=134.209.24.43:1200084494160
[2023-02-02T16:52:51.643008Z INFO sn_node]  
```

**Comment by joshuef** [comment]
I'm not sure this is a bug so much as perhaps poor word choice on the error.

A node can fail to join because it didn't receive an acceptance in <timeout>. Joins may be allowed, so you won't get that error, but the ndoe did not manage to join, so it starts the process over.

Nothing is necessarily being ignored (the elders much reach agreement, and many nodes may be joining, but there's only one space eg). 

So perhaps this needs a reword?

`Failed to join the network in X time. Starting again` ?

The timeout could also be made configurable, 30s in some cases may be too little time.

**Comment by Vort** [comment]
When client makes request, server should respond (if it works correctly).
If server needs more time to create response, it is generally fine _(however, 30s looks too much for me)_.
What is not fine is when response may take time anywhere from 0 seconds to infinity.

I think there should be duration, which is considered as enough time for the answer.
When it is exceeded, this should be treated as bug and fixed.

This is what I expect from properly working software.
Of course, developers may have different opinion and mark timeouts as feature instead of bug.

**Comment by joshuef** [comment]
Perhaps there's a `ConsideringJoinRequest` response to keep alive a node's join efforts. And otherwise, if no joins are allowed we'll just respond with that as we do.

How long that request takes would be another question. There will be a time after which a node may wish to bail though, and that's effectively what we have here, just with an implicit "consideringJoin" as it's not been told to go away...

>  What is not fine is when response may take time anywhere from 0 seconds to infinity.

There are some responses though here as seen in your logs So it seems to me more like an issue of how much feedback is valid and wording rather than an unresponsive server.

As a node operator you may like more feedback on the join process. What feedback should be provided and why? (the vote process is underway... ? Last vote accepted was XYZ? )

----

> This is what I expect from properly working software.
> Of course, developers may have different opinion and mark timeouts as feature instead of bug.

@Vort your bad attitude is leaking from the forum. 

There's no need for sarcasm, snark or condescension. And we'll not bother responding to it if you keep it up. You might believe you weren't doing that, but you're reputation, sadly carries across, and I've no time for it. How you choose to engage is up to you.

**Comment by Vort** [comment]
> As a node operator you may like more feedback on the join process. What feedback should be provided and why?

The most important is to have indication if join process is really happening or network is hung, crashed, or in other way broken.
Right now I see that developers assume that 1 minute is enough _(I wrongly said about 30s previously)_ and when 1min timeout is happening, I make conclusion that network is having problems / bugs.

> There are some responses though here as seen in your logs

Responses were happening for ~250 ms, then ~1 minute (~60000 ms) nothing happened.
Of course, when I read next "in some cases may be too little time", sarcasm start to appear.
Because such behaviour for me is clear sign of hang.

> your bad attitude is leaking from the forum

Quality of this software and approach to its development is the same regardless of where I discuss it.

> And we'll not bother responding to it if you keep it up.

I hope that you will extract useful information from my messages no matter how it will be presented.
Also I hope that with quality of this software going up, amount of my sarcasm will go down.

**Comment by joshuef** [comment]
>  Also I hope that with quality of this software going up, amount of my sarcasm will go down.

That's just it @Vort you don't have to present with any "sarcasm" as you call it. If you want the software to improve, contribute and do so with respect. Sarcasm as you call it, being rude and disrespectful is what I call it. And continuously so.

You may want to read the guidelines on contributing [here](https://github.com/maidsafe/QA/blob/master/CONTRIBUTING.md) and the link to the [contributors coveneant](https://www.contributor-covenant.org/) therein.

But for me your attitude on here and in the forum breaches this quite clearly. As I've mentioned here and above.

@Vort you can decide if acting out is more important to you than working to improve open source projects and your relationships with the people working on them. 

But I'm not having any of more of it. It's a distraction that the team doesn't need.

Should you choose to apologise for your conduct on here, and elsewhere and commit to work on improving it, we can continue to work on this issue and hopefully more down the line. 

Or you can decide that being sarcastic, rude, disrespectful etc is more important to you, and we'll go from _there_.

**Comment by Vort** [comment]
Looks like discussions about discussions is more important for developers, than code by itself.
You had choice to discuss bug (and it would bring constructivity), but you decided to discuss CoCs instead (which will not help to make your code better). That is sad.

> do so with respect

Looks like you think that respect should be present no matter what.
For me respect forms depending on how interactions are going on.
If interactions goes totally wrong, there will be no respect, no matter what CoCs says.
Showing respect when it is absent considered as lie by me and, of course, I don't want to do this.

This project is not that bad to have no respect at all of course.
I see that developers make efforts to have good code and good community.
But because I see lots of problems with both parts, I can't be 100% respectful (as CoC unconditionally demands).

Less problems -> more respect. This is how it should work.

If you decide to ignore disrespectful criticism, you will lose lots of useful information.
Not every user will agree with your preferences of unconditional respect.
But such users will still be able to help.

**Comment by joshuef** [comment]
> but you decided to discuss CoCs instead (which will not help to make your code better). That is sad.

Actually, I discussed the issue and pointed out that I would not tolerate your attitude together. I pointed a way forwards so we might continue without the attitude. But you're not interested in that.

Code is made by people. I want people in this project treated with respect. I don't need to spend my time dealing with poorly behaved people. (All this time I could have been fixing bugs?!)

Even if you don't respect folk or a code base or people running a project, going out of your way to be _disrespectful_ is not on. 

👋
